File: core/src/main/java/com/netflix/iceberg/SnapshotParser.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

package com.netflix.iceberg;

import com.fasterxml.jackson.core.JsonGenerator;
import com.fasterxml.jackson.databind.JsonNode;
import com.google.common.base.Preconditions;
import com.google.common.collect.Lists;
import com.netflix.iceberg.exceptions.RuntimeIOException;
import com.netflix.iceberg.util.JsonUtil;
import java.io.IOException;
import java.io.StringWriter;
import java.util.List;

public class SnapshotParser {

  private static final String SNAPSHOT_ID = "snapshot-id";
  private static final String PARENT_SNAPSHOT_ID = "parent-snapshot-id";
  private static final String TIMESTAMP_MS = "timestamp-ms";
  private static final String MANIFESTS = "manifests";

  static void toJson(Snapshot snapshot, JsonGenerator generator) throws IOException {
    generator.writeStartObject();
    generator.writeNumberField(SNAPSHOT_ID, snapshot.snapshotId());
    if (snapshot.parentId() != null) {
      generator.writeNumberField(PARENT_SNAPSHOT_ID, snapshot.parentId());
    }
    generator.writeNumberField(TIMESTAMP_MS, snapshot.timestampMillis());
    generator.writeArrayFieldStart(MANIFESTS);
    for (ManifestFile file : snapshot.manifests()) {
      generator.writeString(file.path());
    }
    generator.writeEndArray();
    generator.writeEndObject();
  }

  public static String toJson(Snapshot snapshot) {
    try {
      StringWriter writer = new StringWriter();
      JsonGenerator generator = JsonUtil.factory().createGenerator(writer);
      generator.useDefaultPrettyPrinter();
      toJson(snapshot, generator);
      generator.flush();
      return writer.toString();
    } catch (IOException e) {
      throw new RuntimeIOException(e, "Failed to write json for: %s", snapshot);
    }
  }

  static Snapshot fromJson(TableOperations ops, JsonNode node) {
    Preconditions.checkArgument(node.isObject(),
        "Cannot parse table version from a non-object: %s", node);

    long versionId = JsonUtil.getLong(SNAPSHOT_ID, node);
    Long parentId = null;
    if (node.has(PARENT_SNAPSHOT_ID)) {
      parentId = JsonUtil.getLong(PARENT_SNAPSHOT_ID, node);
    }
    long timestamp = JsonUtil.getLong(TIMESTAMP_MS, node);
    List<ManifestFile> manifests = Lists.transform(JsonUtil.getStringList(MANIFESTS, node),
        location -> new GenericManifestFile(location, 0));

    return new BaseSnapshot(ops, versionId, parentId, timestamp, manifests);
  }

  public static Snapshot fromJson(TableOperations ops, String json) {
    try {
      return fromJson(ops, JsonUtil.mapper().readValue(json, JsonNode.class));
    } catch (IOException e) {
      throw new RuntimeIOException(e, "Failed to read version from json: %s", json);
    }
  }
}

File: python/iceberg/core/snapshot_parser.py
# Licensed to the Apache Software Foundation (ASF) under one
# or more contributor license agreements.  See the NOTICE file
# distributed with this work for additional information
# regarding copyright ownership.  The ASF licenses this file
# to you under the Apache License, Version 2.0 (the
# "License"); you may not use this file except in compliance
# with the License.  You may obtain a copy of the License at
#
#   http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing,
# software distributed under the License is distributed on an
# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
# KIND, either express or implied.  See the License for the
# specific language governing permissions and limitations
# under the License.

import json

from .base_snapshot import BaseSnapshot


class SnapshotParser(object):
    SNAPSHOT_ID = "snapshot-id"
    PARENT_SNAPSHOT_ID = "parent-snapshot-id"
    TIMESTAMP_MS = "timestamp-ms"
    SUMMARY = "summary"
    OPERATION = "operation"
    MANIFESTS = "manifests"
    MANIFEST_LIST = "manifest-list"

    @staticmethod
    def to_json(snapshot):
        return json.dumps(SnapshotParser.to_dict(snapshot))

    @staticmethod
    def to_dict(snapshot):
        return {SnapshotParser.SNAPSHOT_ID: snapshot.snapshot_id,
                SnapshotParser.TIMESTAMP_MS: snapshot.timestamp_millis,
                SnapshotParser.PARENT_SNAPSHOT_ID: snapshot.parent_id,
                SnapshotParser.OPERATION: snapshot.operation,
                SnapshotParser.SUMMARY: snapshot.summary,
                SnapshotParser.MANIFESTS: [manifest.manifest_path for manifest in snapshot.manifests]}

    @staticmethod
    def from_json(ops, json_obj, spec=None):
        if isinstance(json_obj, str):
            json_obj = json.loads(json_obj)

        if not isinstance(json_obj, dict):
            raise RuntimeError("Cannot parse table version from a non-object: %s" % json_obj)

        version_id = json_obj.get(SnapshotParser.SNAPSHOT_ID)
        parent_id = json_obj.get(SnapshotParser.PARENT_SNAPSHOT_ID)
        timestamp_millis = json_obj.get(SnapshotParser.TIMESTAMP_MS)
        operation = json_obj.get(SnapshotParser.OPERATION)
        summary = json_obj.get(SnapshotParser.SUMMARY)

        if SnapshotParser.MANIFEST_LIST in json_obj:
            # the manifest list is stored in a manifest list file
            return BaseSnapshot(ops, version_id, parent_id,
                                manifest_list=ops.new_input_file(json_obj.get(SnapshotParser.MANIFEST_LIST)),
                                timestamp_millis=timestamp_millis,
                                operation=operation,
                                summary=summary)
        else:
            manifests = json_obj.get(SnapshotParser.MANIFESTS, list())

            return BaseSnapshot(ops, version_id, parent_id,
                                manifests=manifests,
                                timestamp_millis=timestamp_millis,
                                operation=operation,
                                summary=summary)
